utils/tzif/src/data/tzif.rs

```diff
@@ -297,3 +297,17 @@ pub struct DataBlock {
     /// A series of [`UtLocalIndicator`] objects.
     pub ut_local_indicators: Vec<UtLocalIndicator>,
 }
+
+impl DataBlock {
+    /// Retrieves the timezone designation at index `idx`.
+    pub fn time_zone_designation(&self, mut idx: usize) -> Option<&str> {
+        self.time_zone_designations.iter().find_map(|d| {
+            if idx <= d.len() {
+                Some(&d[idx..])
+            } else {
+                idx -= d.len() + 1;
+                None
+            }
+        })
+    }
+}
```

Comment:
Hi @nordzilla! Thanks for the super detailed review! From reading your description of the behavior of this function in your review I think we're on the same page about how the `idx` of a local time designation works.

As you mention, the `idx` of a local time designation might point anywhere into the concatenation of null terminated timezone designations. Let's take your example of `"LMT\0AEDT\0AEST\0"`. This PR would parse those designations as this list `["LMT", "AEDT", "AEST"]`. Now let's say that we call this function with `idx = 5`, which as you correctly note should map to `"EDT"`.

What this code does is consider each designation in order, checking if `idx` points inside of it (including one past its end). If it does, then we're done and we return `&d[idx..]`. If `idx` points after the end of the currently considered designation then we move to the next one in the list and reduce `idx` by the length of the designation plus one.

Tracing through the specific example where `idx = 5`, we first check if `idx <= "LMT".len()`. This is false, therefore `idx` points to some subsequent designation. We set `idx -= "LMT".len() + 1`, so now `idx = 1` and consider the next designation which is `"AEDT"`. This time `idx <= "AEDT".len()` is true, and so we return `"AEDT[1..]`, which is equal to `"EDT"` as expected.

My PR includes a test that goes over all these edge cases where I check that this function returns the correct designation even when `idx` points to the beginning, middle, end of a designation. You can see the testcases considered here https://github.com/unicode-org/icu4x/pull/5581/files#diff-6050e2a3cc770e7d69a0f364dbff07cda4f734fdc363caef8012875cf6a37425R1153

> The implementation, however, is stuck half way between.
>
> You're still splitting the string up into a Vec<String> delimited by the null-terminator character:

> But your helper function treats the idx value, which is an index into the original byte string with multiple null terminators, instead as an index into the first string in the vector that is long enough for the index.

I hope the explanation above clears this up since the `idx` is not considering only the first string but is correctly moving along all the designations until the correct one is found and then from that returning the correct suffix.

Please let me know if something didn't make sense and I'll happily provide more explanation.

Reply:
You're definitely right! I misunderstood the function in my haste to provide further context in the initial review. I apologize.